scripts/table_updates/utilities.py

```diff
@@ -126,12 +186,95 @@ def revert_table_version(syn, table_id):
+    """
+    # check the validity of bpc_column_list
+    valid_col = column_mapping_table.loc[column_mapping_table["prissmm_form"] == form,].prissmm_element.tolist()
+    assert all(item in valid_col for item in bpc_column_list), (f"Invalid bpc_column_list. Column names should be matching {valid_col}.")
+
+    logger.info(f"Update {bpc_column_list} in {form}")
+    # load bpc table
+    cpt_table_id = master_table.loc[
+        master_table["form"] == form, "id"
+    ].values[0]
+
+    if cohort:
+        condition = f"cohort = '{cohort}'"
+        cpt_dat = download_synapse_table(syn, cpt_table_id, condition = condition)
+    else:
+        cpt_dat = download_synapse_table(syn, cpt_table_id)
+    cpt_dat.index = cpt_dat.index.map(str)
+    cpt_dat["index"] = cpt_dat.index
+    # subset main_genie_table based on bpc_column_list
+    main_genie_column_list = [", ".join(column_mapping_table.loc[column_mapping_table["prissmm_element"] == col,].genie_element) for col in bpc_column_list]
+    
+    if form == "patient_characteristics":
+        main_genie_table = main_genie_table[main_genie_column_list + ["PATIENT_ID"]]
+        cpt_seq_dat = cpt_dat.merge(
+            main_genie_table,
+            how="left",
+            left_on="genie_patient_id",
+            right_on="PATIENT_ID",
+        )
+    else: 
+        main_genie_table = main_genie_table[main_genie_column_list + ["SAMPLE_ID"]]
+        cpt_seq_dat = cpt_dat.merge(
+            main_genie_table,
+            how="left",
+            left_on="cpt_genie_sample_id",
+            right_on="SAMPLE_ID",
+        )
+    cpt_seq_dat.index = cpt_seq_dat["index"]
+    cpt_seq_dat.index.name =  None
+    cpt_seq_dat = cpt_seq_dat[main_genie_column_list]
+    cpt_seq_dat.columns = bpc_column_list
+    # reformat cpt_seq_date column
+    if "cpt_seq_date" in cpt_seq_dat.columns:
+        cpt_seq_dat["cpt_seq_date"] = cpt_seq_dat["cpt_seq_date"].map(float_to_int)
+    return cpt_table_id, cpt_seq_dat
+
+def overwrite_tier1a(syn: synapseclient.Synapse, form: str, cpt_table_id: str, cpt_seq_dat: pandas.DataFrame, bpc_column_list: List[str], logger: logging.Logger = None) -> None:
+    """Function to update tier1a columns with outputs from update_tier1a
+
+    Args:
+        syn (synapseclient.Synapse): The synapse client connection
+        form (str): The form name, can be either patient_characteristics or cancer_panel_test
+        cpt_table_id (str): The synapse ID for BPC table to be modified
+        cpt_seq_dat (pandas.DataFrame): The updated BPC table as dataframe
+        bpc_column_list (List[str]): The column list to be replaced
+        logger (logging.Logger, optional): The custom logger. Defaults to None.
+    """
+    # check the validity of bpc_column_list
+    logger.info(f"Overwrite {bpc_column_list} in {form}")
+    # load bpc table
+    cpt_table_schema = syn.get(cpt_table_id)
+    cpt_dat_query = syn.tableQuery(f"SELECT * FROM {cpt_table_id}")
```

Comment:
Does this need to take into account the cohort?

Reply:
Not necessary. I've tested it. Whether to include condition in the table query or not gives the identical etag. Here is a screenshot of my testing.

<img width="740" alt="Screenshot 2024-10-19 at 12 56 55 AM" src="https://github.com/user-attachments/assets/318a4333-9c36-49f8-9941-14866cf8f847">